teps/0019-other-arch-support.md

```diff
@@ -0,0 +1,424 @@
+
+Extend current Tekton CI/CD system to run Tekton tests and builds for non-amd64
+architectures. As a result Tekton artifacts (container images and yaml files)
+will be tested on other hardware architectures and can be released in the same
+way as it is done for amd64.
+
+Architecture specific builds/tests should be done on real hardware (no emulation
+on top of amd64). That means that native hardware(for specific architecture)
+should be provided by interested parties. Tekton lives on top of k8s cluster,
+so native hardware should provide ability to get new k8s cluster on demand or
+at least have preinstalled k8s cluster(s), which can be cleaned/recreated fast.
+
+It makes no sense to setup full CI/CD infrastructure on native hardware, amd64
+Tekton cluster (dogfooding) has already functionality which can be used as main
+server and architecture specific cluster can be used when it is required
+(actual tests or builds).
+amd64 Tekton cluster will take responsibilities:
+- to initiate all actions.
+- to show results/logs in the UI.
+- to keep all secrets and interact with storages, registries, external systems,
+ etc.
+- to interact with google cloud (google sdk is available only for amd64).
+
+Non-amd64 k8s cluster will:
+- operate on native hardware.
+- build the non-amd64 container images.
+- run the tests.
+
+All steps should be packed as Tekton tasks, that allows to use current
+dogfooding cluster (or similar setup) to build non-amd64 Tekton artifacts via
+Tekton.
+
+### Requirements
+
+amd64 Tekton cluster should be able to:
+- communicate with non-amd64 k8s cluster via kubectl or ssh where it's
+  necessary.
+- have access to non-amd64 local container registry.
+- show the results in the UI
+
+Non-amd64 k8s cluster should have:
+- several worker nodes.
+- local container registry.
+- default storage.
+- access to Internet or required container images stored in local registry.
+
+### Design details
+
+To target some part of the logic to native k8s cluster it is enough to specify
+`kubeconfig` with information about access to non-amd64 cluster, for instance
+`go test ... -kubeconfig ...` to run e2e tests, `export KUBECONFIG ...` and
+`kubectl delete/get ...` for simple kubectl cleanup/setup tasks or `kubectl 
+apply -f *.yaml` for more complex actions.
+Current tasks available at [catalog](https://github.com/tektoncd/catalog/tree/master/task)
+can be used as is ([kubernetes-actions](https://github.com/tektoncd/catalog/tree/master/task/kubernetes-actions/0.1))
+or extended ([kubectl-deploy-pod](https://github.com/tektoncd/catalog/tree/master/task/kubectl-deploy-pod/0.1))
+to work with remote k8s cluster.
+
+### Examples
+
+Examples are presented just to describe main idea and can be modified in the
+future to use tasks from [catalog](https://github.com/tektoncd/catalog/tree/master/task).
+
+#### Example 1: Build architecture specific release for Tekton pipeline
+
+Task is started from amd64 Tekton cluster and runs Tekton pipeline build on
```

Comment:
With recently added support for [building multi-arch images in `ko`](https://github.com/google/ko/pull/38), and [multi-arch support for distroless](https://github.com/GoogleContainerTools/distroless/pull/586) (which we use as a base for most of our images), we can soon start to build multi-arch images using `ko resolve --platform=all`, from the normal dogfooding cluster we use today.

There will need to be some work to make the `git-init` steps, and some others, multi-arch-capable, but until then at least Pipelines that don't use the `git` resource should be supported just fine.

With those aside, once we get Tekton's release onto `--platform=all`, supporting new arch boils down to adding support for that arch to `gcr.io/distroless/static`, which is just a few static files in a container.

Reply:
One thing on `--platform=all`, what if we don't want to support *all* platform ? (aka, I think it does get the list of platform from the base image used)

Reply:
Is there a reason you don't want support for all possible platforms?

If so, we _could) consider modifying `ko` based on our needs to allow us to specify more complex `--platform` flag values. Today, it supports `--platform=amd64` (or any single arch), or `--platform=all` (build for all platforms supported by the base image). If we wanted `--platform=amd64,arm64,s390x` or `--platform=all,-amd64` we could certainly discuss that use case with the `ko` maintainers. But I'd like to make that request with a well-formed reason for why we don't want to build for all available platforms.

Reply:
Yeah, `all` was added as a starting point.  There are a few more "fun" models that @jonjohnsonjr and I were discussing, but decided this was a nice and simple starting point.  Flags is obviously one of the easiest ways to go.

I really wanted to do something smart around intersecting the base images of the collection of import referenced being resolved, but @jonjohnsonjr pointed out that users might want more heterogeneity than this allows (e.g. if I'm targeting a mixed cluster and my node selectors express things my base images do not).

At the end of the day, you have control of the base images you use, so `--platform=all` is a pretty powerful (albeit complex) escape hatch for folks to use here.

Reply:
> Is there a reason you don't want support for all possible platforms?

Yes, man power :upside_down_face:, hardware availability, platform knowledge. We will need to **support** those, fix bugs, reproduce problems, … this is.. some work :upside_down_face:. Currently, `gcr.io/distroless/…` supports amd64 and arm64 right ? so we would support only those (which is fine). What if I want to support more ? Using another base (`ubuntu`) that supports more is fine, but then, maybe `ubuntu` also support `arm` or something I don't have the power to support as of today (hardware availability to test things, etc…).

As long as it supports single arch (`--platform=amd64`), I am fine, we can setup our release pipeline to loop over the arch *we* want to support. `--platform=amd64,arm64,s390x` would be nice to have, but is not something required.

> At the end of the day, you have control of the base images you use, so `--platform=all` is a pretty powerful (albeit complex) escape hatch for folks to use here.

Yes and no :upside_down_face: We can choose the base image, and in that sense, we are in control of the base image. If we choose to use `distroless`, then, we don't control the supported arch anymore (at least not with `all` :upside_down_face:). `--platform=myarch` is good enough for me anyway.

My *only* worry is, what happens if I use `--platform=s390x` and the base image doesn't support it ? :stuck_out_tongue_closed_eyes: We may need to maintain our own set of base image (maybe in `plumbing`) if at some point we want to support arch that are not available in the current base images used :stuck_out_tongue_closed_eyes:

Reply:
btw, I am just thinking outloud, I am really happy `ko` supports multi-platform, and the default behavior of `--platform=all` seems good to me too :stuck_out_tongue: :wink:

Reply:
I was insanely happy when leant that multi-arch support PR was merged to `ko`. And I agree that it makes things much easier. 
At the same time I agree with @vdemeester that `--platform=all` can be too powerful as tests still should be running on native hardware and "support for all possible platforms" for me means that native hardware should be available for these platforms, which is not true at this moment.

Reply:
Technical details :) I verified that `ko resolve --platform=all`, using `alpine` base image in .ko.yaml, builds installable/runnable versions of pipeline, triggers, operator, dashboard for amd64 and s390x (with multi-arch images)

Reply:
I understand the concern about the support burden of having Tekton's images available in a plethora of architectures. If it helps, we can document that Tekton only officially supports `amd64` and `s390x`, for example, even though other architectures might be _available_.

Having this distinction could be helpful, since it means users who might be interested in testing and eventually committing to support `arm64` could play with those out of the box (or, for some other`new-arch`, assuming distroless supports `new-arch`), without having to build their own images.

Reply:
cc @duglin @afrittoli 

IBM had raised a desire to have support for Power and Z in the context of Knative, and a willingness (I think) by IBM to contribute resources for CI.  I thought Tekton might be in a similar boat.

For arm64, I'm going to try to use an EKS Graviton cluster for as much of my development as it will allow.

---
Totally ack the concerns, though.  To Jason's point, perhaps it would be enough to distinguish between "supported" and "best effort"?  You could even produce distinct bundles with `ko resolve --platform={supported}` and `ko resolve --platform=all`.

Feedback on how to control this is certainly welcome and desired 🙏